core/field.js

```diff
@@ -644,39 +669,106 @@ Blockly.Field.prototype.forceRerender = function() {
+ * than this method.
+ * @param {*} newValue New value.
  */
 Blockly.Field.prototype.setValue = function(newValue) {
+  var doLogging = false;
   if (newValue === null) {
-    // No change if null.
+    doLogging && console.log('null, return');
+    // Not a valid value to check.
+    return;
+  }
+
+  newValue = this.doClassValidation_(newValue);
+  if (newValue === null) {
+    doLogging && console.log('invalid, return');
+    this.doValueInvalid_();
+    if (this.isDirty_) {
+      this.forceRerender();
+    }
     return;
   }
-  // Validate input.
-  var validated = this.callValidator(newValue);
-  if (validated !== null) {
-    newValue = validated;
+
+  var localValidator = this.getValidator();
+  if (localValidator) {
+    var validatedValue = localValidator.call(this, newValue);
+    // Sometimes local validators are used as change listeners (bad!) which
+    // means they might return undefined accidentally, so we'll just ignore that.
+    if (validatedValue !== undefined) {
+      newValue = validatedValue;
+    }
+    if (newValue === null) {
+      doLogging && console.log('invalid, return');
+      this.doValueInvalid_();
+      if (this.isDirty_) {
+        this.forceRerender();
+      }
+      return;
+    }
   }
-  // Check for change.
   var oldValue = this.getValue();
-  if (oldValue == newValue) {
+  if (oldValue === newValue) {
+    doLogging && console.log('same, return');
+    // No change.
     return;
   }
+
   if (this.sourceBlock_ && Blockly.Events.isEnabled()) {
     Blockly.Events.fire(new Blockly.Events.BlockChange(
         this.sourceBlock_, 'field', this.name, oldValue, newValue));
   }
-  this.setText(newValue);
+  this.doValueUpdate_(newValue);
+  if (this.isDirty_) {
+    this.forceRerender();
+  }
+  doLogging && console.log(this.value_);
+};
+
+/**
+ * A generic value possessed by the field.
+ * Should generally be non-null, only null when the field is created.
+ * @type {*}
+ * @protected
+ */
+Blockly.Field.prototype.value_ = null;
+
+/**
+ * Used to validate a value. Returns input by default. Can be overridden by
+ * subclasses, see FieldDropdown.
+ * @param {*} newValue The value to be validated.
+ * @return {*} The validated value, same as input by default.
+ * @protected
+ */
+Blockly.Field.prototype.doClassValidation_ = function(newValue) {
+  // For backwards compatibility.
+  newValue = this.classValidator(newValue);
+  return newValue;
+};
+
+/**
+ * Used to update the value of a field. Can be overridden by subclasses to do
+ * custom storage of values/updating of external things.
+ * @param {*} newValue The value to be saved.
+ * @protected
+ */
+Blockly.Field.prototype.doValueUpdate_ = function(newValue) {
+  this.value_ = newValue;
+  this.isDirty_ = true;
+  // For backwards compatibility.
+  this.text_ = String(newValue);
+};
+
+/**
+ * Used to notify the field an invalid value was input. Can be overiden by
+ * subclasses, see FieldTextInput.
+ * No-op by default.
```

Comment:
**With regard to the value:**
The field reverts to the value it had before the start of the "edit". For text fields the edit begins when the user opens the input, and ends when the user closes the input. For a colour field each selection is an individual edit.

This means that [this ](https://user-images.githubusercontent.com/25440652/57254058-2b976400-7005-11e9-9303-f3f16c76a737.gif)colour field /doesn't/ revert to its editor-opened value (red):

But [this ](https://user-images.githubusercontent.com/25440652/57185207-0c20ff80-6e7c-11e9-8bfd-8fb7491b4646.gif)text field /does/ revert to its editor-opened value (default) instead of reverting to a value of aaa.

I didn't realize this was inconsistent until you mentioned it, so this may or may not be what you want. Should be pretty easy to change though =)

**With regard to the text:**
Restated in a more general way, I believe you're asking if by default the display will represent the input value, even if it is invalid? This is not the case, because there is no way for blockly to know what the display is to be able to update it. By default fields will act like the colour field, displaying the last valid value, and having a value of the last valid value. If a developer wants to display invalid values (or put up warnings etc) they will need to override doValueInvalid_

Reply:
We weren't very consistent before and I don't think there's a big difference between the two behaviors. I'm happy with this as is given that this makes it consistent and adds a way to override the default behavior.